From 8746fe773c49fc391dc8eb97eae80e746e780200 Mon Sep 17 00:00:00 2001
From: Varun Sharma <vsharma@habana.ai>
Date: Fri, 29 Apr 2022 16:36:00 +0300
Subject: [PATCH] [SW-84464] greco: fuse: print version

This patch adds support in print fuse version for greco.
Also since it calls efuse_init() in fuse_read(), it takes
care of initialization of efuse, if not already done.
efuse init is required by upper fw layers to use read efuse
values, else it returns 0 values on read.

Change-Id: I6a12b35b97209fd80f1d3b825ec70acd424094ff
Signed-off-by: Varun Sharma <vsharma@habana.ai>
---

diff --git a/arch/greco/common/asic.c b/arch/greco/common/asic.c
index 1f85442..ed1351d 100644
--- a/arch/greco/common/asic.c
+++ b/arch/greco/common/asic.c
@@ -8,15 +8,73 @@
 
 #include <psoc_global_conf_regs.h>
 #include <cpu_if_regs.h>
-
+#include <string.h>
 #include <bitops.h>
 #include <preboot.h>
 #include <uart.h>
 #include <pll.h>
 #include <hl_boot_if.h>
 #include <pci.h>
+#include <efuse.h>
 #include <address_space.h>
 
+#define GRECO_FUSE_TABLE_SIZE	4
+#define GRECO_FUSE_DEVID_SIZE	8
+#define GRECO_FUSE_FAB_SIZE	2
+#define GRECO_FUSE_LOT_SIZE	6
+#define GRECO_FUSE_WAFER_SIZE	2
+#define GRECO_FUSE_XADDR_SIZE	2
+#define GRECO_FUSE_YADDR_SIZE	2
+
+struct hl_fuse_data {
+	union {
+		struct {
+			char table_ver[GRECO_FUSE_TABLE_SIZE];
+			char dev_id[GRECO_FUSE_DEVID_SIZE];
+			char mem_config;
+			char fab[GRECO_FUSE_FAB_SIZE];
+			char process_type;
+			char lot[GRECO_FUSE_LOT_SIZE];
+			char wafer[GRECO_FUSE_WAFER_SIZE];
+			char xaddr[GRECO_FUSE_XADDR_SIZE];
+			char yaddr[GRECO_FUSE_YADDR_SIZE];
+		} __packed;
+		u32 data[7];
+	};
+};
+
+int hl_fuse_get_version(char *version, size_t size)
+{
+	struct hl_fuse_data fuse;
+	struct preboot_fuse_ver_data ver_data = {
+		.table_ver = fuse.table_ver,
+		.table_ver_sz = sizeof(fuse.table_ver),
+		.dev_id = fuse.dev_id,
+		.dev_id_sz = sizeof(fuse.dev_id),
+		.fab = fuse.fab,
+		.fab_sz = sizeof(fuse.fab),
+		.lot = fuse.lot,
+		.lot_sz = sizeof(fuse.lot),
+		.wafer = fuse.wafer,
+		.wafer_sz = sizeof(fuse.wafer),
+		.xaddr = fuse.xaddr,
+		.xaddr_sz = sizeof(fuse.xaddr),
+		.yaddr = fuse.yaddr,
+		.yaddr_sz = sizeof(fuse.yaddr),
+	};
+	int i;
+
+	for (i = 0; i < ARRAY_SIZE(fuse.data); i++)
+		if (preboot_fuse_read(i, &fuse.data[i]))
+			return 1;
+
+	memcpy(hl_asic_sn, fuse.dev_id, sizeof(hl_asic_sn));
+
+	preboot_fuse_version_serialize(version, size, &ver_data, true);
+
+	return 0;
+}
+
 u32 hl_skip_pcie_init(void)
 {
 	/* Use DumpSel[2] to skip PCIe initialization */
